capsules/src/virtual_aes_ccm.rs

```diff
@@ -0,0 +1,798 @@
+use core::cell::Cell;
+use kernel::common::cells::{OptionalCell, TakeCell};
+use kernel::common::{List, ListLink, ListNode};
+// use kernel::debug;
+use kernel::hil::symmetric_encryption;
+use kernel::hil::symmetric_encryption::{
+    AES128Ctr, AES128, AES128CBC, AES128_BLOCK_SIZE, AES128_KEY_SIZE, CCM_NONCE_LENGTH,
+};
+use kernel::ReturnCode;
+
+// the CCM state is not public is aes_ccm.rs
+#[derive(Copy, Clone, Eq, PartialEq, Debug)]
+enum CCMState {
+    Idle,
+    Auth,
+    Encrypt,
+}
+
+// to cache up the function parameters of the crypt() function
+struct CryptFunctionParameters {
+    buf: &'static mut [u8],
+    a_off: usize,
+    m_off: usize,
+    m_len: usize,
+    mic_len: usize,
+    confidential: bool,
+    encrypting: bool,
+}
+
+impl CryptFunctionParameters {
+    pub fn new(
+        buf: &'static mut [u8],
+        a_off: usize,
+        m_off: usize,
+        m_len: usize,
+        mic_len: usize,
+        confidential: bool,
+        encrypting: bool,
+    ) -> CryptFunctionParameters {
+        CryptFunctionParameters {
+            buf: buf,
+            a_off: a_off,
+            m_off: m_off,
+            m_len: m_len,
+            mic_len: mic_len,
+            confidential: confidential,
+            encrypting: encrypting,
+        }
+    }
+}
+
+pub struct MuxAES128CCM<'a, A: AES128<'a> + AES128Ctr + AES128CBC> {
+    aes: &'a A,
+    clients: List<'a, VirtualAES128CCM<'a, A>>,
+    inflight: OptionalCell<&'a VirtualAES128CCM<'a, A>>,
+    // debug_counter: Cell<i32>,
+}
+
+impl<'a, A: AES128<'a> + AES128Ctr + AES128CBC> MuxAES128CCM<'a, A> {
+    pub fn new(aes: &'a A) -> MuxAES128CCM<'a, A> {
+        // aes.enable(); // enable the hardware
+        // debug!("new mux");
+        MuxAES128CCM {
+            aes: aes,
+            clients: List::new(),
+            inflight: OptionalCell::empty(),
+            // debug_counter: Cell::new(0),
+        }
+    }
+
+    /// manually re-enable the hardware
+    pub fn enable(&self) {
+        self.aes.enable();
+    }
+
+    /// disable the underlying hardware
+    pub fn disable(&self) {
+        self.aes.disable();
+    }
+
+    fn do_next_op(&self) {
+        if self.inflight.is_none() {
+            let mnode = self.clients.iter().find(|node| node.queued_up.is_some());
+            mnode.map(|node| {
+                self.inflight.set(node);
+                let parameters: CryptFunctionParameters = node.queued_up.take().unwrap();
+                let (res, _) = node.crypt_r(
+                    parameters.buf,
+                    parameters.a_off,
+                    parameters.m_off,
+                    parameters.m_len,
+                    parameters.mic_len,
+                    parameters.confidential,
+                    parameters.encrypting,
+                );
+                // notice that we didn't put the parameters back...
+
+                if res != ReturnCode::SUCCESS {
+                    // notify the client that there's a failure
```

Comment:
Do we want to indicate error/debug log if `buf` or `client` is None?

Reply:
added checks for buf / client being None in recent commit.